Source/Core/VideoCommon/NetPlayChatUI.cpp

```diff
@@ -0,0 +1,97 @@
+// Copyright 2019 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "VideoCommon/NetPlayChatUI.h"
+
+#include <imgui.h>
+
+constexpr float DEFAULT_WINDOW_WIDTH = 220.0f;
+constexpr float DEFAULT_WINDOW_HEIGHT = 400.0f;
+
+constexpr size_t MAX_BACKLOG_SIZE = 100;
+
+std::unique_ptr<NetPlayChatUI> g_netplay_chat_ui;
+
+NetPlayChatUI::NetPlayChatUI(std::function<void(const std::string&)> callback)
+{
+  message_callback = std::move(callback);
+}
+
+void NetPlayChatUI::Display()
+{
+  const float scale = ImGui::GetIO().DisplayFramebufferScale.x;
+
+  ImGui::SetNextWindowPos(ImVec2(10.0f * scale, 10.0f * scale), ImGuiCond_FirstUseEver);
+  ImGui::SetNextWindowSizeConstraints(
+      ImVec2(DEFAULT_WINDOW_WIDTH * scale, DEFAULT_WINDOW_HEIGHT * scale),
+      ImGui::GetIO().DisplaySize);
+  if (!ImGui::Begin("Chat", nullptr, ImGuiWindowFlags_None))
+  {
+    ImGui::End();
+    return;
+  }
+
+  ImGui::BeginChild("Scrolling", ImVec2(0, -30 * scale), true, ImGuiWindowFlags_None);
+  for (const auto& msg : messages)
+  {
+    auto c = msg.second;
+    ImGui::PushTextWrapPos(0.0f);
+    ImGui::TextColored(ImVec4(c[0], c[1], c[2], 1.0f), "%s", msg.first.c_str());
+    ImGui::PopTextWrapPos();
+  }
+
+  if (scroll_to_bottom)
+  {
+    ImGui::SetScrollHere(1.0f);
+    scroll_to_bottom = false;
+  }
+
+  is_scrolled_to_bottom = ImGui::GetScrollY() == ImGui::GetScrollMaxY();
+
+  ImGui::EndChild();
+
+  ImGui::Spacing();
+
+  ImGui::PushItemWidth(-50.0f * scale);
+  ImGui::InputText("", message_buf, IM_ARRAYSIZE(message_buf));
+  ImGui::PopItemWidth();
+
+  ImGui::SameLine();
+
+  if (ImGui::Button("Send"))
+  {
+    if (message_callback != nullptr)
```

Comment:
This should also check that the string isn't empty, so you can't send blank messages.